AssemblerLib/VectorMatrixAssembler.h

```diff
@@ -49,15 +57,41 @@ class VectorMatrixAssembler
     {
         assert(_data_pos.size() > id);
 
-        LocalToGlobalIndexMap::RowColumnIndices const& indices = _data_pos[id];
+        std::vector<GlobalIndexType> indices;
+
+        // Local matrices and vectors will always be ordered by component,
+        // no matter what the order of the global matrix is.
+        for (unsigned c=0; c<_data_pos.getNumComponents(); ++c)
+        {
+            auto const& idcs = _data_pos(id, c).rows;
+            indices.reserve(indices.size() + idcs.size());
+            indices.insert(indices.end(), idcs.begin(), idcs.end());
+        }
+
+        std::vector<double> localX;
+        std::vector<double> localX_pts;
+
+        if (_x)         localX.reserve(indices.size());
+        if (_x_prev_ts) localX_pts.reserve(indices.size());
+
+        for (auto i : indices)
+        {
+            if (_x)         localX.emplace_back(_x->get(i));
```

Comment:
In PETScVector, get(PestcInt idx) calls PETSc routine
```
VecGetValues(_v, 1, &idx, &x);
```
e.g. it get one entry through an array with the size of one.
To fetch values from petsc vector, communication operation is involved among node. To get a bunch of values through an array in one time costs much less time than to get it one by one.